resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/KubernetesClusterSchedulerBackend.scala

```diff
@@ -0,0 +1,427 @@
+    .get(KUBERNETES_DRIVER_POD_NAME)
+    .getOrElse(throw new SparkException("Must specify the driver pod name"))
+  private implicit val requestExecutorContext = ExecutionContext.fromExecutorService(
+    requestExecutorsService)
+
+  private val driverPod = kubernetesClient.pods()
+    .inNamespace(kubernetesNamespace)
+    .withName(kubernetesDriverPodName)
+    .get()
+
+  override val minRegisteredRatio =
+    if (conf.getOption("spark.scheduler.minRegisteredResourcesRatio").isEmpty) {
+      0.8
+    } else {
+      super.minRegisteredRatio
+    }
+
+  private val executorWatchResource = new AtomicReference[Closeable]
+  protected val totalExpectedExecutors = new AtomicInteger(0)
+
+  private val driverUrl = RpcEndpointAddress(
+    conf.get("spark.driver.host"),
+    conf.getInt("spark.driver.port", DEFAULT_DRIVER_PORT),
+    CoarseGrainedSchedulerBackend.ENDPOINT_NAME).toString
+
+  private val initialExecutors = SchedulerBackendUtils.getInitialTargetExecutorNumber(conf)
+
+  private val podAllocationInterval = conf.get(KUBERNETES_ALLOCATION_BATCH_DELAY)
+
+  private val podAllocationSize = conf.get(KUBERNETES_ALLOCATION_BATCH_SIZE)
+
+  private val allocatorRunnable = new Runnable {
+
+    // Maintains a map of executor id to count of checks performed to learn the loss reason
+    // for an executor.
+    private val executorReasonCheckAttemptCounts = new mutable.HashMap[String, Int]
+
+    override def run(): Unit = {
+      handleDisconnectedExecutors()
+      val executorsToAllocate = mutable.Map[String, Pod]()
+      val currentTotalRegisteredExecutors = totalRegisteredExecutors.get
+      val currentTotalExpectedExecutors = totalExpectedExecutors.get
+      val currentNodeToLocalTaskCount = getNodesWithLocalTaskCounts
+      if (currentTotalRegisteredExecutors < runningExecutorsToPods.size) {
+        logDebug("Waiting for pending executors before scaling")
+      } else if (currentTotalExpectedExecutors <= runningExecutorsToPods.size) {
+        logDebug("Maximum allowed executor limit reached. Not scaling up further.")
+      } else {
+        for (i <- 0 until math.min(
+          currentTotalExpectedExecutors - runningExecutorsToPods.size, podAllocationSize)) {
+          val executorId = EXECUTOR_ID_COUNTER.incrementAndGet().toString
+          val executorPod = executorPodFactory.createExecutorPod(
+            executorId,
+            applicationId(),
+            driverUrl,
+            conf.getExecutorEnv,
+            driverPod,
+            currentNodeToLocalTaskCount)
+          require(executorPod.getMetadata.getLabels.containsKey(SPARK_EXECUTOR_ID_LABEL),
+            s"Illegal internal state for pod with name ${executorPod.getMetadata.getName} - all" +
+              s" executor pods must contain the label $SPARK_EXECUTOR_ID_LABEL.")
+          val resolvedExecutorIdLabel = executorPod.getMetadata.getLabels.get(
+            SPARK_EXECUTOR_ID_LABEL)
+          require(resolvedExecutorIdLabel == executorId,
```

Comment:
What's the behavior you want when these checks trigger?

Right now this will just stop this periodic task (which is the the behavior of `ScheduledExecutorService` when a task throws an exception). It won't stop the application or do anything else, so you'll probably end up with a stuck app or some other weird state.

Reply:
I don't really see a possibility for a successfully constructed executor pod to not have this label with the right value. There's already checks in `ExecutorPodFactory` that guard against users setting this label. So I don't think the checks here are really necessary. @foxish to confirm.

Reply:
+1 to @liyinan926's comment. Since we're directly getting the pod from `ExecutorPodFactory`, the guard is superfluous and basically dead code. @mccheah PTAL.

Reply:
Removed the checks.

Reply:
I disagree with removing the checks here. The contract that the labels are present are in ExecutorPodFactory, which is further removed from this class. It would be good for the contract to be explicit here, because we can then catch errors in integration tests and diagnose them quickly.

Reply:
If I understand correctly, the checks should never fail in real cases. The code creates `executorId`, passes it into `executorPodFactory.createExecutorPod`, which then sets the `SPARK_EXECUTOR_ID_LABEL` to the `executorId` passed in. Now after the `createExecutorPod` call, we check the presence and value correctness of the label on the returned executor pod. This seems to me like doing a unit test on the label here.

Reply:
The checks should never fail unless the code changes in the future, and changing the code in ExecutorPodFactory as well as the ExecutorPodFactorySuite would break the assumptions the scheduler backend is making here. The scheduler backend tests wouldn't fail since we mock the ExecutorPodFactory in those tests.

Reply:
It's the future code change that these requirements are meant to protect against such that the integration tests fail.

Reply:
If this is purely for preventing future changes from potentially breaking it, why cannot we verify that the contract is held in integration tests so the integration tests will fail if the contract is indeed broken?

Reply:
It depends I suppose on how clearly and loudly the integration tests will fail if the contract is not satisfied. I like having the `require` checks here because we fail fast and explicitly.

Reply:
I agree that this depends on how clearly integration tests fail on this. But I still don't think this is a good place for such kind of checks. Also as @vanzin pointed out, failing the checks here only stops the scheduled task without stopping the driver/app so it really does not fail fast nor explicitly in real cases.

Reply:
Fair enough. I'd like to consider how we can make the contract explicit and more connected, but this is fine for now.